src/type-data-frame.c

```diff
@@ -174,6 +174,153 @@ static SEXP c_data_frame_class(SEXP cls) {
   return chr_c(cls, classes_data_frame);
 }
 
+
+SEXP data_frame(SEXP x, r_ssize size, const struct name_repair_opts* p_name_repair_opts);
+
+// [[ register() ]]
+SEXP vctrs_data_frame(SEXP x, SEXP size, SEXP name_repair) {
+  struct name_repair_opts name_repair_opts = new_name_repair_opts(name_repair, args_empty, false);
+  PROTECT_NAME_REPAIR_OPTS(&name_repair_opts);
+
+  r_ssize c_size = 0;
+  if (size == R_NilValue) {
+    c_size = vec_size_common(x, 0);
+  } else {
+    c_size = size_validate(size, ".size");
+  }
+
+  SEXP out = data_frame(x, c_size, &name_repair_opts);
+
+  UNPROTECT(1);
+  return out;
+}
+
+static SEXP df_splice(SEXP x, r_ssize n);
+
+SEXP data_frame(SEXP x, r_ssize size, const struct name_repair_opts* p_name_repair_opts) {
+  if (TYPEOF(x) != VECSXP) {
+    Rf_errorcall(R_NilValue, "Internal error: `x` must be a list.");
+  }
+
+  x = PROTECT(vec_recycle_common(x, size));
+
+  r_ssize n_cols = r_length(x);
+
+  // Unnamed columns are auto-named with `""`
+  if (r_names(x) == R_NilValue) {
+    r_poke_names(x, r_new_character(n_cols));
+  }
+
+  x = PROTECT(df_splice(x, n_cols));
+
+  SEXP names = PROTECT(r_names(x));
+  names = vec_as_names(names, p_name_repair_opts);
+  r_poke_names(x, names);
+
+  SEXP out = new_data_frame(x, size);
+
+  UNPROTECT(3);
+  return out;
+}
+
+static SEXP df_splice(SEXP x, r_ssize n) {
+  SEXP names = PROTECT(r_names(x));
+  const SEXP* p_names = STRING_PTR_RO(names);
+
+  bool any_needs_splice = false;
+  r_ssize i = 0;
+
+  for (; i < n; ++i) {
+    // Only splice unnamed data frames
+    if (p_names[i] != strings_empty) {
+      continue;
+    }
+
+    SEXP col = VECTOR_ELT(x, i);
+
+    if (is_data_frame(col)) {
+      any_needs_splice = true;
+      break;
+    }
+  }
```

Comment:
Can we get rid of the first loop and compute the width right away, with shortcut if the width is unchanged?

Reply:
This does short circuit if the width is unchanged. If we get through the first loop and don't find any unnamed data frames then we are done (i.e. the width is unchanged)

Reply:
I'm proposing to get rid of the first loop and short-cirtuiting after the second loop if width is unchanged. Reduces code complexity?

Reply:
Oh, the point of the first loop is to never allocate the `splice` vector in the common case of having no data frame inputs to splice. It does add a little complexity, but I think it is worth it here to avoid the allocation